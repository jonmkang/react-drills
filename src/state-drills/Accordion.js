import React from 'react';

class Accordion extends React.Component {
    static defaultProps = {
        sections: []
    };

    state = {
        currentTabIndex: -1
    };

    handleButtonClick = (index) => {
        console.log('Button was clicked', index)
        this.setState({ currentTabIndex: index })
    }

    renderSections() {
        if(this.props.sections.length){
            return this.props.sections.map((section, index) => (
                <li>
                    <button key={index} onClick={() => this.handleButtonClick(index)}>
                        {section.title}
                    </button>
                    {this.renderContent(index)}
                </li>
            ))
        }else{
            return (<li></li>)
        }
        
    }

    renderContent(index){
        const currentTab = this.state.currentTabIndex;
        if(currentTab === index){
            return (
                <p>
                    {!!(currentTab === index) && this.props.sections[index].content}
                </p>
            )
        }else{
            return
        }
    }

    render(){
        return (
            <ul>
                {this.renderSections()}
            </ul>
        )
    }

}

export default Accordion;