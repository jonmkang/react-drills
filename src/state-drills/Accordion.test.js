import React from 'react';
import ReactDOM from 'react-dom';
import Accordion from './Accordion';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Accordion Component', () =>{
    const sectionsProp = [
        {
          title: 'Section 1',
          content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
        },
        {
          title: 'Section 2',
          content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
        },
        {
          title: 'Section 3',
          content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
        },
    ];

    it('renders without errors', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Accordion />, div)
        ReactDOM.unmountComponentAtNode(div);
    })

    it('renders no sections by default', () => {
        const tree = renderer.create(<Accordion sections={sectionsProp} />).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('renders an empty li when prop is not supplied', () => {
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders a clicked section properly', () => {
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(0).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders the last section clicked properly', () => {
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(2).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

})