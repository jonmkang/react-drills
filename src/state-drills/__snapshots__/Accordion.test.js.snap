// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component renders a clicked section properly 1`] = `
<ul>
  <li>
    <button
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li>
    <button
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li>
    <button
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders an empty li when prop is not supplied 1`] = `
<ul>
  <li />
</ul>
`;

exports[`Accordion Component renders no sections by default 1`] = `
<ul>
  <li>
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li>
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li>
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders the last section clicked properly 1`] = `
<ul>
  <li>
    <button
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li>
    <button
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li>
    <button
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
    <p>
      Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
    </p>
  </li>
</ul>
`;
